L1Trigger/L1TZDC/plugins/L1TZDCProducer.cc

```diff
@@ -0,0 +1,237 @@
+  const int peakBX = 4;
+  const int bxRPlusMinus = 2;
+  const int startBXR = peakBX - bxRPlusMinus;
+  const int endBXR = peakBX + bxRPlusMinus;
+
+  //rawadc[detector index][time slices]
+  unsigned short rawadc[18][10];
+
+  // the loop below loops over all the elements of the QIE10DigiCollection. Each entry corresponds to one channel
+  for (QIE10DigiCollection::const_iterator it = zdcDigiCollection->begin(); it != zdcDigiCollection->end(); it++) {
+    const QIE10DataFrame& frame(*it);
+    HcalZDCDetId cell = frame.id();
+    int zside = cell.zside();
+    int section = cell.section();
+    int channel = cell.channel();
+
+    if (zside != -1 && zside != 1)
+      continue;
+    if (section != 1 && section != 2)
+      continue;
+    if (section == 1 && (channel < 1 || channel > 5))
+      continue;
+    if (section == 2 && (channel < 1 || channel > 4))
+      continue;
+
+    int ihitid = (zside == 1 ? 9 : 0) + (section == 2 ? 5 : 0) + (channel - 1);
+    //the loop below iterates over the time slices
+    for (int iTS = 0; iTS < nSamples; iTS++) {
+      unsigned short adc = (unsigned short)frame[iTS].adc();
+      rawadc[ihitid][iTS] = adc;
+    }  // end of loop over iTS
+  }    //end of loop over channels
+
+  for (int ibx = 0; ibx < nSamples; ibx++) {
+    double cEMP = 0, cEMM = 0, cHDP = 0, cHDM = 0;
+    double sumcEMP = 0, sumcEMM = 0, sumcHDP = 0, sumcHDM = 0;
+    //idet=0-4 correpond to the EM channels
+    for (int idet = 0; idet < 5; idet++) {
+      unsigned short EMP = rawadc[idet + 9][ibx];
+      unsigned short EMM = rawadc[idet][ibx];
+
+      int cEMP_LUTIndex = zdcLUTIndexHelper(idet + 9, (int)EMP);
+      int cEMM_LUTIndex = zdcLUTIndexHelper(idet, (int)EMM);
+
+      cEMP = ((double)m_params->zdcLUT()->data(cEMP_LUTIndex)) / ((double)m_scaleFactor);
+      cEMM = ((double)m_params->zdcLUT()->data(cEMM_LUTIndex)) / ((double)m_scaleFactor);
+
+      sumcEMP = sumcEMP + cEMP;
+      sumcEMM = sumcEMM + cEMM;
+    }
+    //idet=5-8 correspond to HAD channels
+    for (int idet = 5; idet < 9; idet++) {
+      unsigned short HDP = rawadc[idet + 9][ibx];
+      unsigned short HDM = rawadc[idet][ibx];
+
+      int cHDP_LUTIndex = zdcLUTIndexHelper(idet + 9, (int)HDP);
+      int cHDM_LUTIndex = zdcLUTIndexHelper(idet, (int)HDM);
+
+      cHDP = ((double)m_params->zdcLUT()->data(cHDP_LUTIndex)) / ((double)m_scaleFactor);
+      cHDM = ((double)m_params->zdcLUT()->data(cHDM_LUTIndex)) / ((double)m_scaleFactor);
+
+      sumcHDP = sumcHDP + cHDP;
+      sumcHDM = sumcHDM + cHDM;
+    }
+    double sumM = sumcEMM + sumcHDM;
+    double sumP = sumcEMP + sumcHDP;
+
+    if (ibx == 4) {
+      edm::LogInfo("L1TZDCProducer") << ", sumM= " << sumM << std::endl;
+      edm::LogInfo("L1TZDCProducer") << ", sumP= " << sumP << std::endl;
+    }
+    l1t::EtSum tempEtM = l1t::EtSum();
+    tempEtM.setHwPt(sumM);
+    tempEtM.setHwEta(-1.);
+    tempEtM.setHwPhi(0.);
+    tempEtM.setType(EtSum::EtSumType::kZDCM);
```

Comment:
on a longer term I would suggest to add a constructor to the `l1t::EtSum` object that can take these values directly, instead of using the various "set" methods.